scipp-imaging-2D.ipynb

```diff
@@ -224,218 +162,516 @@
+    "pycharm": {
+     "name": "#%%\n"
+    }
    },
    "outputs": [],
    "source": [
-    "# Adds a coordinate for the spectra\n",
-    "ds.coords[Dim.Spectrum] = sc.Variable([Dim.Spectrum], values=np.arange(ds[\"sample\"].shape[1]))\n",
-    "ds"
+    "integrated = sc.Dataset()\n",
+    "for dim in stitched.coords.keys():\n",
+    "    integrated.coords[dim] = stitched.coords[dim]\n",
+    "\n",
+    "integrated = sc.sum(stitched, 'tof')\n",
+    "# Put TOF coords back for the export to TIFF stack later\n",
+    "integrated.coords[\"tof\"] = stitched.coords[\"tof\"]"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
+   "outputs": [],
+   "source": [
+    "# Pull out reference from out dataset to avoid checking for it in loops\n",
+    "reference = stitched[\"reference\"]\n",
+    "integrated_reference = integrated[\"reference\"]\n",
+    "\n",
+    "# Causes segmentation fault, see scipp/scipp issue: #1176\n",
+    "# when this can be added back we can remove the \"if \"reference\" checks later\n",
+    "# del stitched[\"reference\"]\n",
+    "# del integrated[\"reference\"]\n"
+   ],
    "metadata": {
-    "scrolled": true
-   },
+    "collapsed": false,
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
    "outputs": [],
    "source": [
-    "stitched = sc.Dataset(coords={Dim.Tof: sc.Variable([Dim.Tof], values=np.arange(*rebin_parameters, dtype=np.float64))})\n",
-    "\n",
-    "stitched[\"sample\"] = stitch(ds[\"sample\"], frame_parameters, rebin_parameters)\n",
-    "stitched[\"reference\"] = stitch(ds[\"reference\"], frame_parameters, rebin_parameters)\n",
-    "\n",
-    "stitched"
-   ]
+    "num_spectra = stitched.coords[\"spectrum\"].shape[0]\n",
+    "bank_width = int(np.sqrt(num_spectra))\n",
+    "\n",
+    "def mask_non_sample_regions(integrated_dataset, stitched_dataset, var_name):\n",
+    "    integrated_spectra = integrated_dataset[var_name]\n",
+    "\n",
+    "    # This should use sc.greater once #1178 is completed\n",
+    "    spectra_masks = np.greater(integrated_spectra.values, masking_threshold)\n",
+    "    spectra_masks = sc.Variable([\"spectrum\"], values=spectra_masks)\n",
+    "\n",
+    "    # Some pixels may be noisy / not strong enough signal so they erranously\n",
+    "    # become unmasked / masked from the above operation. This step checks if *all*\n",
+    "    # the surrounding pixels are True or False then sets the center pixel to\n",
+    "    # that value if they are, removing \"salt and pepper\" noise\n",
+    "    final_mask = operations.mask_from_adj_pixels(spectra_masks, bank_width)\n",
+    "    stitched_dataset.masks[var_name] = final_mask\n",
+    "\n",
+    "# Calculate transmission value for all\n",
+    "integrated /= integrated_reference\n",
+    "\n",
+    "for k in integrated.keys():\n",
+    "    if k == \"reference\":  # This can be removed once #1176 is resolved\n",
+    "        continue\n",
```

Comment:
Can you explain the problem? It feels like something is backwards if you are actually affected by this.

Reply:
Once we normalize by open beam (`stitched / reference`) there is no point doing anything else with reference

However because we can't `del d["reference"]` we are forced to pay for something we don't need if we don't check for it.
If the original issue is unresolvable we could rewrite the start to keep `reference` out of the original dataset. But I'd rather keep this in as IMO it shows a user facing problem that needs to be resolved

Reply:
> But I'd rather keep this in as IMO it shows a user facing problem that needs to be resolved

Are you referring to the bug? There is an issue for this.

The problem here is that it is very hard to understand what you are doing:

- Why do you need to add something to `d` then remove it again?
- Why do you then have to skip it?
- Why do you talk about dividing by itself above?

Isn't all that solved by just not putting it into `d`? Or if you need it there initially, just remove it, so you don't have to mess around later, just don't store a reference to it.

Reply:
> just remove it, so you don't have to mess around later, just don't store a reference to it.

But I feel like this is showing a scipp problem, I want to 

1. Do the same processing steps on all items in the dataset (including reference)
2. Take reference out of the dataset but hold onto a reference so
2a) Divide all items in the dataset by reference
3. Del reference as I no longer need it from this point onwards

2a. Is the part which is "hard" since I can't do something naive like
```
ref = d["ref"]
del d["ref"]

d / ref
```

Reply:
```python
ref = d["ref"].copy()
del d["ref"]
d / ref
```
Will eventually add `extract`, so you can do
```python
ref = d.extract('ref')
```
but that is further down the imaginary feature list.